Applications/Spire/Source/Ui/FilteredDropDownMenu.cpp

```diff
@@ -0,0 +1,214 @@
+#include "Spire/Ui/FilteredDropDownMenu.hpp"
+#include <QKeyEvent>
+#include <QPainter>
+#include "Spire/Spire/Dimensions.hpp"
+
+using namespace Spire;
+
+namespace {
+  auto PADDING() {
+    static auto padding = scale_width(8);
+    return padding;
+  }
+}
+
+FilteredDropDownMenu::FilteredDropDownMenu(std::vector<QVariant> items,
+    QWidget* parent)
+    : TextInputWidget(parent),
+      m_was_click_focused(false) {
+  setAttribute(Qt::WA_Hover);
+  setFocusPolicy(Qt::StrongFocus);
+  connect(this, &QLineEdit::editingFinished, this,
+    &FilteredDropDownMenu::on_editing_finished);
+  connect(this, &QLineEdit::textEdited, this,
+    &FilteredDropDownMenu::on_text_edited);
+  if(!items.empty()) {
+    m_current_item = items.front();
+    setText(m_item_delegate.displayText(m_current_item));
+  }
+  m_menu_list = new DropDownList({}, false, this);
+  m_item_activated_connection = m_menu_list->connect_activated_signal(
+    [=] (const auto& value) { on_item_activated(value); });
+  m_list_selection_connection = m_menu_list->connect_value_selected_signal(
+    [=] (const auto& value) { on_item_selected(value, true); });
+  set_items(std::move(items));
+  m_menu_list->installEventFilter(this);
+}
+
+bool FilteredDropDownMenu::eventFilter(QObject* watched, QEvent* event) {
+  if(watched == m_menu_list) {
+    if(event->type() == QEvent::Show) {
+      m_menu_list->setFixedWidth(width());
+      update();
+    }
+  }
+  return TextInputWidget::eventFilter(watched, event);
+}
+
+void FilteredDropDownMenu::focusInEvent(QFocusEvent* event) {
+  TextInputWidget::focusInEvent(event);
+  if(event->reason() == Qt::FocusReason::MouseFocusReason) {
+    m_was_click_focused = true;
+  }
+}
+
+void FilteredDropDownMenu::focusOutEvent(QFocusEvent* event) {
+  if(!m_menu_list->isActiveWindow()) {
+    TextInputWidget::focusOutEvent(event);
+  }
+}
+
+void FilteredDropDownMenu::keyPressEvent(QKeyEvent* event) {
+  switch(event->key()) {
+    case Qt::Key_Delete:
+      setText("");
+      return;
+    case Qt::Key_Escape:
+      setText(m_item_delegate.displayText(m_current_item));
```

Comment:
Is it the case that pressing ESC performs everything that pressing ENTER performs?